util/topgen/secure_prng.py

```diff
@@ -0,0 +1,309 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""Generating random variables by using a cryptographically secure PRNG.
+
+A class for generating random numbers and permutations.
+This class is based on a cryptographically secure PRNG CTR_DRBG described in
+NIST SP 800-90A, section 10.2.1
+https://csrc.nist.gov/publications/detail/sp/800-90a/rev-1/final
+This PRNG guarantees forward and backward secrecy (PRNG output cannot be
+computed from any previous or future outputs) as well as enhanced backward
+secrecy (previous PRNG outputs cannot be computed from a compromised
+internal state).
+"""
+
+import logging as log
+import sys
+from Crypto.Cipher import AES
+from math import ceil as _ceil
+from math import log as _log
+
+
+class secure_prng():
+
+    blocklen = 128
+    keylen = 128
+    # Specification parameters (for easy comparison, not used in the code):
+    #     seedlen = blocklen + keylen = 256
+    #     ctr_len = blocklen = 128
+
+    # Maximal number of requests before reseed is required
+    RESEED_MAX = 2 ** 48
+
+    def __init__(self):
+        """Initialize an instance.
+
+        Default state value is 0.
+        """
+
+        # A 128 bit PRNG output
+        self.returned_bits = []
+
+        # CTR_DRBG Internal state as defined in NIST SP 800-90A, page 50
+        # The state consists of:
+        #     - The value V of 128 bits
+        #     - The value Key of 128 bits
+        #     - Reseed counter
+        # Administrative information such as security_strength and
+        # prediction_resistance_flag are missing in this implementation
+        # because they are not used by the consuming application.
+        self.V = None
+        self.Key = None
+        self.reseed_counter = None
+
+    def CTR_DRBG_Update(self, provided_data_0, provided_data_1):
+        """The update function
+
+        CTR_DRBG_Update process
+        Described in NIST SP 800-90A, page 51
+        For convenience, input provided_data is devided in two 128 bit chunks.
+        """
+        cipher = AES.new(self.Key.to_bytes(16, 'big'), AES.MODE_ECB)
+
+        # Steps 1 and 2
+        # Loop in step 2 consists of only 2 iterations, it is unrolled here
+        # Note ctr_len = blocklen = 128, Therefore we don't go through steps
+        # 2.1.1. and 2.1.2
+        pt0 = self.V + 1
+        pt1 = self.V + 2
```

Comment:
These need to wrap around the block length:
```suggestion
        pt0 = (self.V + 1) % (1 << self.blocklen)
        pt1 = (self.V + 2) % (1 << self.blocklen)
```